File: methods/root.py
from math import *
import numpy as np

def f(x):
  return np.exp(x) - x - 1;

def df(x):
  return np.exp(x) - 1;

def d2f(x):
  return np.exp(x);

def root(f,df,d2f,x0,tol,n):
  xant = x0;
  fant = f(xant);
  e_abs=1000;
  iteration = 1;
  
  while iteration<=n:
    xact = xant - fant * df(xant) / ((df(xant))**2 - fant * d2f(xant));
    fact = f(xact);
    e_abs = abs(xact-xant);
    
    print("Iteration:", iteration, "  X =", xact, "F(x) =", fact, "Error =", e_abs);
    
    iteration += 1;
    xant = xact;
    fant = fact;
    
    if e_abs<tol:
      print("Solution found in X =", xact, "     iterations:", iteration-1, "    error =", e_abs);
      break;
  
  if iteration > n:
    print("Solution not found for tolerance = ", tol);
    
root(f,df,d2f,1,10**-7,100);

File: methods/matlab-methods/root.m
%This program finds the solution to the equation f(x)=0 using the method of multiple roots

%Inputs:
%f, continuous function
%f', continuous function
%f'', continuous function
%x0, initial approximation
%tol, tolerance
%Nmax, maximum number of iterations

%Output:
%x, solution
%iter, number of iterations
%err, error

function [x,iter,err]=C7_raicesmlt(f,df,d2f,x0,tol,Nmax)

    %initialization
    xant=x0;
    fant=f(xant);
    E=1000; 
    cont=0;
    
    %Loop
    while E>tol && cont<Nmax
      xact=xant-fant*df(xant)/((df(xant))^2-fant*d2f(xant));
      fact=f(xact);
      E=abs(xact-xant);
      cont=cont+1;
      xant=xact;
      fant=fact;
    end
    
    %Delivery of results
    x=xact;
    iter=cont;
    err=E;
    end